extensions/parquet/table/src/main/java/io/deephaven/parquet/table/pagestore/OffsetIndexBasedColumnChunkPageStore.java

```diff
@@ -0,0 +1,154 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.parquet.table.pagestore;
+
+import io.deephaven.base.verify.Assert;
+import io.deephaven.base.verify.Require;
+import io.deephaven.chunk.attributes.Any;
+import io.deephaven.engine.page.ChunkPage;
+import io.deephaven.parquet.table.pagestore.topage.ToPage;
+import io.deephaven.parquet.base.ColumnChunkReader;
+import io.deephaven.parquet.base.ColumnPageReader;
+import org.apache.parquet.internal.column.columnindex.OffsetIndex;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.lang.ref.WeakReference;
+import java.util.Arrays;
+
+final class OffsetIndexBasedColumnChunkPageStore<ATTR extends Any> extends ColumnChunkPageStore<ATTR> {
+    private final OffsetIndex offsetIndex;
+    private final int numPages;
+    /**
+     * Set if first ({@link #numPages}-1) pages have equal number of rows
+     */
+    private boolean isPageSizeFixed;
+    /**
+     * Fixed number of rows per page, only valid if {@link #isPageSizeFixed} is true. Used to map from row index to page
+     * number.
+     */
+    private final long fixedPageSize;
+    private final Object[] objectsForSynchronizingPageAccess;
+    private final ColumnPageReader[] columnPageReaders;
+    private final WeakReference<PageCache.IntrusivePage<ATTR>>[] pages;
+
+    OffsetIndexBasedColumnChunkPageStore(@NotNull final PageCache<ATTR> pageCache,
+            @NotNull final ColumnChunkReader columnChunkReader,
+            final long mask,
+            @NotNull final ToPage<ATTR, ?> toPage) throws IOException {
+        super(pageCache, columnChunkReader, mask, toPage);
+        offsetIndex = columnChunkReader.getOffsetIndex();
+        Assert.assertion(offsetIndex != null, "offsetIndex != null");
+        numPages = offsetIndex.getPageCount();
+
+        // noinspection unchecked
+        pages = (WeakReference<PageCache.IntrusivePage<ATTR>>[]) new WeakReference[numPages];
+        columnPageReaders = new ColumnPageReader[numPages];
+
+        isPageSizeFixed = true;
+        final long firstPageSize;
+        if (numPages > 1) {
+            firstPageSize = offsetIndex.getFirstRowIndex(1) - offsetIndex.getFirstRowIndex(0);
+        } else {
+            firstPageSize = numRows();
+        }
+        objectsForSynchronizingPageAccess = new Object[numPages];
+        for (int i = 0; i < numPages; ++i) {
+            objectsForSynchronizingPageAccess[i] = new Object();
```

Comment:
Instead of using an array of locks (which is what this is), we could probably race to construct readers and CAS them into place in an `AtomicReferenceArray`, then synchronize on the readers as the previous implementations did.

To clarify, the patterns you might consider look like:
```
localReader = array.get(index);
if (localReader == null && !array.compareAndSet(index, null, localReader = constructReader(index)) {
    localReader = array.get(index);
} 
```
You might be able to simplify:
```
localReader = array.get(index);
if (localReader == null) {
    localReader = array.compareAndGet(index, e -> e == null ? constructReader(index) : e);
} 
```

Reply:
The pattern we have kept in the final patch is similar to how `updateAndGet` is internally implemented using `weakCompareAndSetVolatile` inside a loop. 
Thanks to Devin for suggesting that approach.